django/thunderstore/social/providers.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import requests
+from django.conf import settings
+
+
+class BaseOauthHelper:
+    token = None
+
+    def __init__(self, code: str, redirect_url: str) -> None:
+        self.code = code
+        self.redirect_url = redirect_url
+
+    def complete_login(self) -> None:
+        """
+        Fetch access token with the temporary code and configured secret.
+        """
+        data = {
+            "client_id": self.CLIENT_ID,
+            "client_secret": self.CLIENT_SECRET,
+            "code": self.code,
+            "grant_type": "authorization_code",  # Required by Discord
+            "redirect_uri": self.redirect_url,  # Required by Discord
+        }
+        headers = {"Accept": "application/json"}
+
+        response = requests.post(self.OAUTH_URL, data, headers=headers)
+        response_data = response.json()
+        self.token = response_data["access_token"]
+
+    def get_user_info(self) -> Dict:
```

Comment:
As mentioned in the class definition comment, there should be some "good" way to type-safely declare abstract methods that are required to be implemented by sub-classes. Could be worth exploring.